### PR TITLE
A jessesweirdbranch

### PR DESCRIPTION
**Pull request by JesseStephenBrown**
Here's the weird shit. I will try to write up what I have done and why in the diff near the loc

### CODE REVIEW

File: Entry.cs [hunk 1]
```diff
@@ -20,10 +20,10 @@ public class Entry {
     public KINDS Kind {get; private set;}
     public int Size {get; private set;}
     public int Offset {get; set;}
-    public List<string> Parameters {get; private set;}
+    public List<Parameter> Parameters {get; private set;}
```
Comment (JesseStephenBrown) [inline Entry.cs:23]: We need to have mode and lexeme

Comment (JesseStephenBrown) [inline Entry.cs:23]: Actually, no we fuggin' don't. We just need mode.

File: SemAnalyzer.cs [hunk 2]
```diff
@@ -58,6 +58,12 @@ public class SemAnalyzer{
         );
     }
 
+    public void genAddSP(int count){
+        output(
+            "ADD SP #" + count + " SP"
+        );
+    }
+
```
Comment (JesseStephenBrown) [inline SemAnalyzer.cs:66]: This can be used when making room for the register, or the local vars.

File: parser.cs [hunk 3]
```diff
@@ -205,6 +206,7 @@ public class Parser {
                 TYPES typeRec = type();
                 SymbolTable top = __symbolTableStack.Peek();
                 if(typeRec != TYPES.NONE) {
+                    top.IncSize(); // Make room for PC
```
Comment (JesseStephenBrown) [inline parser.cs:209]: When we start processing vars, we need to leave an empty offset between the params and the vars

File: parser.cs [hunk 4]
```diff
@@ -266,7 +268,6 @@ public class Parser {
     private void procedureDeclaration() {
         switch(__lookahead.Type) {
             case TOKENS.PROCEDURE:
-                string label =
```
Comment (JesseStephenBrown) [inline parser.cs:269]: We're going to need a label, obviously this commit is mid-workflow

File: parser.cs [hunk 5]
```diff
@@ -320,7 +321,7 @@ public class Parser {
                     new SymbolTable(
                         identifier,
                         __symbolTableStack.Peek().NestingLevel +1,
-                        0,
+                        1,
```
Comment (JesseStephenBrown) [inline parser.cs:324]: Offsets start at 1. This allows room for the old register value on the stack

File: parser.cs [hunk 6]
```diff
@@ -1011,14 +1012,9 @@ public class Parser {
         switch(__lookahead.Type) {
             case TOKENS.IDENTIFIER:
                 Console.Write(67 + " ");
+                __analyzer.genAddSP(1);
```
Comment (JesseStephenBrown) [inline parser.cs:1015]: make room for the register

File: parser.cs [hunk 7]
```diff
@@ -1469,13 +1466,22 @@ public class Parser {
...
                 KINDS identifierKind = __symbolTableStack.Peek().GetKind(identifier);
-                if(identifierKind == KINDS.VAR || identifierKind == KINDS.PARAMETER) {
+                if(identifierKind == KINDS.VAR) {
                     Console.Write(116 + " ");
                     String factorId = variableIdentifier();
                     TYPES factorType = __symbolTableStack.Peek().GetType(factorId);
                     factorRec = new SemRecord(factorType, factorId);
                     __analyzer.genPushVar(factorRec);
                     return factorRec;
+                } else if (identifierKind == KINDS.PARAMETER) {
```
Comment (JesseStephenBrown) [inline parser.cs:1476]: Here I started work, but it needs completion. Basically, what we need to know is what the mode is of the particular entry so we know if we need to push say, 1(D0) or @1(D0)

File: Entry.cs [hunk 8]
```diff
@@ -32,4 +32,23 @@ public class Entry {
...
+    }
+}
+
+public class Parameter {
+    public string Lexeme {get; private set;}
+    public boolean InOut {get; private set;}
+
+    public Parameter(string lexeme, inOut=True){
+        Lexeme = lexeme;
+        InOut  = inOut;
```
Comment (brianpmaher) [inline Entry.cs:52]: in == true and out == false?

**Comment by brianpmaher** [comment]
Could you update scrumdo tickets so I know exactly where we're at on this? I'd like to take it over from here, but not sure what is "completed" and what is "in progress" without spending a while going through each one.

**Pull request closed without merging**